# Analisis Sentimen Ulasan Aplikasi Shopee di google play store Menggunakan Metode Klasifikasi Algoritma Naive Bayes

# scraping data

In [1]:
#download library google-play-scraper
!pip install google-play-scraper

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.2/50.2 kB 2.3 MB/s eta 0:00:00


# Impor paket yang diperlukan

In [2]:
from google_play_scraper import app

import pandas as pd

import numpy as np

In [3]:
#scrape jumlah ulasan yang diinginkan
from google_play_scraper import Sort, reviews

result, continuation_token = reviews(
    'com.shopee.id',
    lang='id',  #disini kita mau men scrape data ulasan aplikasi shopee yang berada di google play store
    country='id', #kita setting bahasa nya menjadi bahasa indonesia
    sort=Sort.MOST_RELEVANT, # # kemudian kita gunakan most_relevan untuk mendapatkan ulasan yang paling relevant
    count=1000, # disini jumlah ulasan yang mau kita ambil ada seribu
    filter_score_with=None # # kemudian di filter_score kita gunakan None untuk mengambil semua score atau ratting bintang 1 sampai 5
)

In [4]:
df_busu = pd.DataFrame(np.array(result),columns=['review'])

df_busu = df_busu.join(pd.DataFrame(df_busu.pop('review').tolist()))

df_busu.head()

reviewId       userName  \
0  f0d74060-9fdb-48a9-b2f7-0fe7a2f29707  Sariati waode   
1  25caf768-5463-46a1-92ad-227da4b4981f   Cepi Cahyana   
2  6eb967ea-b7b3-4b14-8b82-7450fd2e0214  Wahyu Saputra   
3  8072bb2b-1ce7-4371-bbd2-a7a2109b409c  Ainun Nadiroh   
4  d63e95de-a4d6-4efc-8c12-33302ab81821    Ika Adillah   

                                           userImage  \
0  https://play-lh.googleusercontent.com/a-/ALV-U...   
1  https://play-lh.googleusercontent.com/a-/ALV-U...   
2  https://play-lh.googleusercontent.com/a-/ALV-U...   
3  https://play-lh.googleusercontent.com/a/ACg8oc...   
4  https://play-lh.googleusercontent.com/a-/ALV-U...   

                                             content  score  thumbsUpCount  \
0  Saya suka dengan shoope, tapi tolong untuk eks...      3             89   
1  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5            478   
2  Sekarang lemot banget, respon lama, tombol bac...      1            579   
3  Sekarang pengirimannya lama banget. Sebelum2ny...      3             17   
4  shopee sekarang meskipun ada vocher kaya nggad...      2             32   

  reviewCreatedVersion                  at  \
0              3.33.50 2024-09-16 22:51:30   
1              3.33.50 2024-09-17 11:28:53   
2              3.33.50 2024-09-04 02:51:36   
3              3.34.37 2024-09-18 08:31:34   
4              3.33.50 2024-09-17 09:28:11   

                                        replyContent           repliedAt  \
0  Hi kak, mohon maaf ya kak atas ketidaknyamanan... 2024-09-16 23:31:31   
1  Hai kak, maaf ya terkait kendala kamu. terimak... 2024-09-17 12:21:24   
2  Hai kak, maaf ya bikin cemas. terkait kendala ... 2022-09-26 18:03:11   
3  Hai kak, maaf untuk kendala terkait expedisi. ... 2024-09-18 09:33:27   
4  Hai, Kak. Maaf ya buat kamu gak nyaman. Terkai... 2022-09-15 12:57:03   

  appVersion  
0    3.33.50  
1    3.33.50  
2    3.33.50  
3    3.34.37  
4    3.33.50

In [5]:
len(df_busu.index) #kemudian hitung kembali berapa jumlah data yg didapatkan

1000

In [6]:
df_busu[['userName', 'score','at', 'content']].head()  #dari scrapping tsb didapatkan banyak sekali kolom, kemudian kolom" tsb kita filter
                                                        #sehingga didapatkan kolom username, score, at dan content

userName  score                  at  \
0  Sariati waode      3 2024-09-16 22:51:30   
1   Cepi Cahyana      5 2024-09-17 11:28:53   
2  Wahyu Saputra      1 2024-09-04 02:51:36   
3  Ainun Nadiroh      3 2024-09-18 08:31:34   
4    Ika Adillah      2 2024-09-17 09:28:11   

                                             content  
0  Saya suka dengan shoope, tapi tolong untuk eks...  
1  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...  
2  Sekarang lemot banget, respon lama, tombol bac...  
3  Sekarang pengirimannya lama banget. Sebelum2ny...  
4  shopee sekarang meskipun ada vocher kaya nggad...

In [7]:
#Run This Code to Sort the Data By Date

new_df = df_busu[['userName', 'score','at', 'content']]
sorted_df = new_df.sort_values(by='at', ascending=False) #Sort by Newst, change to True if you want to sort by Oldest.
sorted_df.head()

userName  score                  at  \
871  nini pahriani      1 2024-09-18 17:51:00   
240      Hadi Guna      1 2024-09-18 17:10:22   
322      Saya Dary      1 2024-09-18 16:34:49   
431      Kang Pono      1 2024-09-18 15:46:09   
974   Fitalis Seko      4 2024-09-18 14:54:38   

                                               content  
871  duh..apa lagi ini ..mkin d upgrade .mkin parah...  
240  Kecewa makin lama Pelayanan buruk apa lgi jasa...  
322  suka banget dengan upgrade pembatalan pembelia...  
431  shopee makin lama aja proses pengirimannya. Du...  
974  Sedikit kesal, karna g pesan barang malah ada ...

In [8]:
my_df = sorted_df[['userName', 'score','at', 'content']] #kemudian kita simpan ke variabel my_df

In [9]:
my_df=my_df[['content', 'score']]#karena kita hanya membutuhkan kolom content dan score maka kita lakukan filter kolom lgi hingga menyisakan kolom content dan score.


In [10]:
my_df.head()

content  score
871  duh..apa lagi ini ..mkin d upgrade .mkin parah...      1
240  Kecewa makin lama Pelayanan buruk apa lgi jasa...      1
322  suka banget dengan upgrade pembatalan pembelia...      1
431  shopee makin lama aja proses pengirimannya. Du...      1
974  Sedikit kesal, karna g pesan barang malah ada ...      4

# PELABELAN

In [11]:
def pelabelan(score):
  if score < 3:
    return 'Negatif'
  elif score == 4 :
    return 'Positif'
  elif score == 5 :
    return 'Positif'
my_df['Label'] = my_df ['score'].apply(pelabelan)
my_df.head(50)

content  score    Label
871  duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif
240  Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif
322  suka banget dengan upgrade pembatalan pembelia...      1  Negatif
431  shopee makin lama aja proses pengirimannya. Du...      1  Negatif
974  Sedikit kesal, karna g pesan barang malah ada ...      4  Positif
38   Saya kecewa selama ini download toko online la...      5  Positif
621  Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif
124  Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif
268  Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif
21   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif
372  Ga tau knp kurir SPX sekarang nganterin nya la...      3     None
409  Lelet bgt sekarang anjir,mau cari satu produk ...      3     None
37   Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif
3    Sekarang pengirimannya lama banget. Sebelum2ny...      3     None
961  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif
8    Notifikasi dikembalikan ke semula saja yang ad...      5  Positif
18   Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif
19   Saya member platinum shopee, tp akhir akhir in...      5  Positif
10   TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif
25   Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif
216  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif
11   ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif
341  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif
65   Belanja online ya cm di shopee...sangat mudah ...      5  Positif
266  Saya kasih 🌟 dulu, soalnya spaylater & spinjam...      1  Negatif
12   Sistem pengiriman lebih sering memakan waktu l...      2  Negatif
52   Aplikasi penipuan, baru 1 jam check out, karen...      1  Negatif
105  Awalnya bagus tapi makin kesini kok malah di b...      3     None
14   Saya sangat puas belanja di shoope,, barang se...      5  Positif
5    Shopee tambah tidak bagus dulu bisa milih jasa...      1  Negatif
28   sekarang di shopee banyak penipu, penjual tida...      1  Negatif
703  Apa guna nya jadi member platinum kalo paket m...      1  Negatif
6    Makin kesini makin berat nih apk nya, terjadi ...      1  Negatif
913  Terima kasih Shopee, kasus permasalahan saya s...      4  Positif
42   Bayar belum jatuh tempo sudah di bayar ( dan s...      1  Negatif
1    SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5  Positif
29   Shopee emang bagus dari awal, namun semenjak a...      3     None
77   Update terbaru sudah! Jaringan full 4G lte! Cl...      1  Negatif
16   Tolong kembalikan kebebasan memilih jasa kirim...      1  Negatif
4    shopee sekarang meskipun ada vocher kaya nggad...      2  Negatif
744  Makin lama makin parah makin pelit, biaya admi...      1  Negatif
253  Katanya terus belanja pakai spaylater supaya b...      1  Negatif
51   Minta batal pesanan ditolak sama toko. Barang ...      1  Negatif
32   Hallo pihak shopee, saya suka belanja online m...      4  Positif
125  Halo shopee...untuk pemilihan ekspedisi itu ke...      3     None
466  Belanja di shopee banyak variasi produknya, ku...      5  Positif
7    Bisa ngga iklannya ga sampe mengganggu orang g...      1  Negatif
169  kembalikan fitur yg bisa ganti ekspedisi soaln...      1  Negatif
450  Masyallah suka sekali dengan aplikasi ini seja...      5  Positif
686  tolong dibantu ya pihak shopee, saya setiap pe...      2  Negatif

In [12]:
my_df.to_csv("scrapped_data.csv", index = False)  #kemudian save menjadi file csv

# pembersihan data
# data cleaning

In [13]:
import pandas as pd
pd.set_option('display.max_columns', None)
my_df = pd.read_csv('/content/scrapped_data.csv')
my_df.head(50)

content  score    Label
0   duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif
1   Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif
2   suka banget dengan upgrade pembatalan pembelia...      1  Negatif
3   shopee makin lama aja proses pengirimannya. Du...      1  Negatif
4   Sedikit kesal, karna g pesan barang malah ada ...      4  Positif
5   Saya kecewa selama ini download toko online la...      5  Positif
6   Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif
7   Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif
8   Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif
9   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif
10  Ga tau knp kurir SPX sekarang nganterin nya la...      3      NaN
11  Lelet bgt sekarang anjir,mau cari satu produk ...      3      NaN
12  Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif
13  Sekarang pengirimannya lama banget. Sebelum2ny...      3      NaN
14  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif
15  Notifikasi dikembalikan ke semula saja yang ad...      5  Positif
16  Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif
17  Saya member platinum shopee, tp akhir akhir in...      5  Positif
18  TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif
19  Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif
20  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif
21  ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif
22  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif
23  Belanja online ya cm di shopee...sangat mudah ...      5  Positif
24  Saya kasih 🌟 dulu, soalnya spaylater & spinjam...      1  Negatif
25  Sistem pengiriman lebih sering memakan waktu l...      2  Negatif
26  Aplikasi penipuan, baru 1 jam check out, karen...      1  Negatif
27  Awalnya bagus tapi makin kesini kok malah di b...      3      NaN
28  Saya sangat puas belanja di shoope,, barang se...      5  Positif
29  Shopee tambah tidak bagus dulu bisa milih jasa...      1  Negatif
30  sekarang di shopee banyak penipu, penjual tida...      1  Negatif
31  Apa guna nya jadi member platinum kalo paket m...      1  Negatif
32  Makin kesini makin berat nih apk nya, terjadi ...      1  Negatif
33  Terima kasih Shopee, kasus permasalahan saya s...      4  Positif
34  Bayar belum jatuh tempo sudah di bayar ( dan s...      1  Negatif
35  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5  Positif
36  Shopee emang bagus dari awal, namun semenjak a...      3      NaN
37  Update terbaru sudah! Jaringan full 4G lte! Cl...      1  Negatif
38  Tolong kembalikan kebebasan memilih jasa kirim...      1  Negatif
39  shopee sekarang meskipun ada vocher kaya nggad...      2  Negatif
40  Makin lama makin parah makin pelit, biaya admi...      1  Negatif
41  Katanya terus belanja pakai spaylater supaya b...      1  Negatif
42  Minta batal pesanan ditolak sama toko. Barang ...      1  Negatif
43  Hallo pihak shopee, saya suka belanja online m...      4  Positif
44  Halo shopee...untuk pemilihan ekspedisi itu ke...      3      NaN
45  Belanja di shopee banyak variasi produknya, ku...      5  Positif
46  Bisa ngga iklannya ga sampe mengganggu orang g...      1  Negatif
47  kembalikan fitur yg bisa ganti ekspedisi soaln...      1  Negatif
48  Masyallah suka sekali dengan aplikasi ini seja...      5  Positif
49  tolong dibantu ya pihak shopee, saya setiap pe...      2  Negatif

In [14]:
# info() digunakan untuk menampilkan informasi detail tentang dataframe,
#seperti jumlah baris data, nama-nama kolom berserta jumlah data dan tipe datanya, dan sebagainya.
my_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   content  1000 non-null   object
 1   score    1000 non-null   int64 
 2   Label    854 non-null    object
dtypes: int64(1), object(2)
memory usage: 23.6+ KB


In [15]:
#Tampilkan setiap baris yang memiliki nilai null (NaN) pada kolom apapun
#Gunakan fitur isna() yang disediakan library pandas
my_df.isna()

content  score  Label
0      False  False  False
1      False  False  False
2      False  False  False
3      False  False  False
4      False  False  False
..       ...    ...    ...
995    False  False  False
996    False  False  False
997    False  False  False
998    False  False  False
999    False  False  False

[1000 rows x 3 columns]

In [16]:
my_df.isna().any()

content    False
score      False
Label       True
dtype: bool

In [17]:
my_df.describe()

score
count  1000.000000
mean      2.526000
std       1.616148
min       1.000000
25%       1.000000
50%       2.000000
75%       4.000000
max       5.000000

In [18]:
#mencari jumlah baris data yang bernilai null
#terdapat kolom label memiliki nilai kosong
my_df.isnull().sum()

content      0
score        0
Label      146
dtype: int64

# 1. Handling Missing value-Ignore tuple


In [19]:
my_df.dropna(subset=['Label'],inplace = True)

In [20]:
my_df.isnull().sum()

content    0
score      0
Label      0
dtype: int64

In [21]:
my_df.head(50)

content  score    Label
0   duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif
1   Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif
2   suka banget dengan upgrade pembatalan pembelia...      1  Negatif
3   shopee makin lama aja proses pengirimannya. Du...      1  Negatif
4   Sedikit kesal, karna g pesan barang malah ada ...      4  Positif
5   Saya kecewa selama ini download toko online la...      5  Positif
6   Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif
7   Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif
8   Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif
9   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif
12  Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif
14  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif
15  Notifikasi dikembalikan ke semula saja yang ad...      5  Positif
16  Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif
17  Saya member platinum shopee, tp akhir akhir in...      5  Positif
18  TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif
19  Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif
20  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif
21  ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif
22  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif
23  Belanja online ya cm di shopee...sangat mudah ...      5  Positif
24  Saya kasih 🌟 dulu, soalnya spaylater & spinjam...      1  Negatif
25  Sistem pengiriman lebih sering memakan waktu l...      2  Negatif
26  Aplikasi penipuan, baru 1 jam check out, karen...      1  Negatif
28  Saya sangat puas belanja di shoope,, barang se...      5  Positif
29  Shopee tambah tidak bagus dulu bisa milih jasa...      1  Negatif
30  sekarang di shopee banyak penipu, penjual tida...      1  Negatif
31  Apa guna nya jadi member platinum kalo paket m...      1  Negatif
32  Makin kesini makin berat nih apk nya, terjadi ...      1  Negatif
33  Terima kasih Shopee, kasus permasalahan saya s...      4  Positif
34  Bayar belum jatuh tempo sudah di bayar ( dan s...      1  Negatif
35  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5  Positif
37  Update terbaru sudah! Jaringan full 4G lte! Cl...      1  Negatif
38  Tolong kembalikan kebebasan memilih jasa kirim...      1  Negatif
39  shopee sekarang meskipun ada vocher kaya nggad...      2  Negatif
40  Makin lama makin parah makin pelit, biaya admi...      1  Negatif
41  Katanya terus belanja pakai spaylater supaya b...      1  Negatif
42  Minta batal pesanan ditolak sama toko. Barang ...      1  Negatif
43  Hallo pihak shopee, saya suka belanja online m...      4  Positif
45  Belanja di shopee banyak variasi produknya, ku...      5  Positif
46  Bisa ngga iklannya ga sampe mengganggu orang g...      1  Negatif
47  kembalikan fitur yg bisa ganti ekspedisi soaln...      1  Negatif
48  Masyallah suka sekali dengan aplikasi ini seja...      5  Positif
49  tolong dibantu ya pihak shopee, saya setiap pe...      2  Negatif
51  Kenapa Shopee sudah semakin berat loadingnya l...      4  Positif
52  Jadi sekarang shopee lebih buruk dari yang lai...      1  Negatif
54  Awal dulu bintang 5 trs krn suatu hal yg menur...      5  Positif
55  Dulu saya kasih bintang 5 karena emang memperm...      1  Negatif
56  Bingung aku sama shope skrng jasa kirim nya pa...      1  Negatif
57  TOP bangeet .. Udah lama menggunakan aplikasi ...      5  Positif

In [22]:
my_df.to_csv("shopeepreprocessing.csv", index = False)  #simpan hasil file data cleaning dengan nama shopeepreprocessing.csv

# Text PreProcessing

In [23]:
import pandas as pd
df = pd.read_csv('/content/shopeepreprocessing.csv')
df.head(50)

content  score    Label
0   duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif
1   Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif
2   suka banget dengan upgrade pembatalan pembelia...      1  Negatif
3   shopee makin lama aja proses pengirimannya. Du...      1  Negatif
4   Sedikit kesal, karna g pesan barang malah ada ...      4  Positif
5   Saya kecewa selama ini download toko online la...      5  Positif
6   Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif
7   Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif
8   Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif
9   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif
10  Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif
11  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif
12  Notifikasi dikembalikan ke semula saja yang ad...      5  Positif
13  Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif
14  Saya member platinum shopee, tp akhir akhir in...      5  Positif
15  TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif
16  Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif
17  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif
18  ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif
19  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif
20  Belanja online ya cm di shopee...sangat mudah ...      5  Positif
21  Saya kasih 🌟 dulu, soalnya spaylater & spinjam...      1  Negatif
22  Sistem pengiriman lebih sering memakan waktu l...      2  Negatif
23  Aplikasi penipuan, baru 1 jam check out, karen...      1  Negatif
24  Saya sangat puas belanja di shoope,, barang se...      5  Positif
25  Shopee tambah tidak bagus dulu bisa milih jasa...      1  Negatif
26  sekarang di shopee banyak penipu, penjual tida...      1  Negatif
27  Apa guna nya jadi member platinum kalo paket m...      1  Negatif
28  Makin kesini makin berat nih apk nya, terjadi ...      1  Negatif
29  Terima kasih Shopee, kasus permasalahan saya s...      4  Positif
30  Bayar belum jatuh tempo sudah di bayar ( dan s...      1  Negatif
31  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5  Positif
32  Update terbaru sudah! Jaringan full 4G lte! Cl...      1  Negatif
33  Tolong kembalikan kebebasan memilih jasa kirim...      1  Negatif
34  shopee sekarang meskipun ada vocher kaya nggad...      2  Negatif
35  Makin lama makin parah makin pelit, biaya admi...      1  Negatif
36  Katanya terus belanja pakai spaylater supaya b...      1  Negatif
37  Minta batal pesanan ditolak sama toko. Barang ...      1  Negatif
38  Hallo pihak shopee, saya suka belanja online m...      4  Positif
39  Belanja di shopee banyak variasi produknya, ku...      5  Positif
40  Bisa ngga iklannya ga sampe mengganggu orang g...      1  Negatif
41  kembalikan fitur yg bisa ganti ekspedisi soaln...      1  Negatif
42  Masyallah suka sekali dengan aplikasi ini seja...      5  Positif
43  tolong dibantu ya pihak shopee, saya setiap pe...      2  Negatif
44  Kenapa Shopee sudah semakin berat loadingnya l...      4  Positif
45  Jadi sekarang shopee lebih buruk dari yang lai...      1  Negatif
46  Awal dulu bintang 5 trs krn suatu hal yg menur...      5  Positif
47  Dulu saya kasih bintang 5 karena emang memperm...      1  Negatif
48  Bingung aku sama shope skrng jasa kirim nya pa...      1  Negatif
49  TOP bangeet .. Udah lama menggunakan aplikasi ...      5  Positif

# Case Folding
Proses case folding adalah proses mengubah seluruh huruf menjadi huruf kecil. Pada proses ini karakter-karakter 'A'-'Z' yang terdapat pada data diubah kedalam karakter 'a'-'z'.

In [24]:
import re
def  clean_text(df, text_field, new_text_field_name):
    my_df[new_text_field_name] = my_df[text_field].str.lower()
    my_df[new_text_field_name] = my_df[new_text_field_name].apply(lambda elem: re.sub(r"(@[A-Za-z0-9]+)|([^0-9A-Za-z \t])|(\w+:\/\/\S+)|^rt|http.+?", "", elem))
    # remove numbers
    my_df[new_text_field_name] = my_df[new_text_field_name].apply(lambda elem: re.sub(r"\d+", "", elem))
    return my_df

In [25]:
my_df['text_clean'] = my_df['content'].str.lower()
my_df['text_clean']
data_clean = clean_text(my_df, 'content', 'text_clean')
data_clean.head(10)


content  score    Label  \
0  duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif   
1  Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif   
2  suka banget dengan upgrade pembatalan pembelia...      1  Negatif   
3  shopee makin lama aja proses pengirimannya. Du...      1  Negatif   
4  Sedikit kesal, karna g pesan barang malah ada ...      4  Positif   
5  Saya kecewa selama ini download toko online la...      5  Positif   
6  Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif   
7  Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif   
8  Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif   
9  Saya baru tau ada toko megic ataupun penipuan,...      5  Positif   

                                          text_clean  
0  duhapa lagi ini mkin d upgrade mkin parah pela...  
1  kecewa makin lama pelayanan buruk apa lgi jasa...  
2  suka banget dengan upgrade pembatalan pembelia...  
3  shopee makin lama aja proses pengirimannya dul...  
4  sedikit kesal karna g pesan barang malah ada p...  
5  saya kecewa selama ini download toko online la...  
6  asli aplikasi lemot banget mau ganti nomor aja...  
7  kalo aku kasih bintang  pasti ga dibaca kalo t...  
8  aplikasi shopee sangat memberikan kemudahan ba...  
9  saya baru tau ada toko megic ataupun penipuan ...

# Stopword Removal
Stopword adalah kata umum yang biasanya muncul dalam jumlah besar dan dianggap tidak memiliki makna. Contoh stopword dalam bahasa Indonesia adalah “yang”, “dan”, “di”, “dari”, dll. Makna di balik penggunaan stopword yaitu dengan menghapus kata-kata yang memiliki informasi rendah dari sebuah teks, kita dapat fokus pada kata-kata penting sebagai gantinya.

In [26]:
import nltk.corpus
nltk.download('stopwords')
from nltk.corpus import stopwords
stop = stopwords.words('indonesian')
data_clean['text_StopWord'] = data_clean['text_clean'].apply(lambda x:' '.join([word for word in x.split() if word not in (stop)]))
data_clean.head(50)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


content  score    Label  \
0   duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif   
1   Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif   
2   suka banget dengan upgrade pembatalan pembelia...      1  Negatif   
3   shopee makin lama aja proses pengirimannya. Du...      1  Negatif   
4   Sedikit kesal, karna g pesan barang malah ada ...      4  Positif   
5   Saya kecewa selama ini download toko online la...      5  Positif   
6   Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif   
7   Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif   
8   Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif   
9   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif   
12  Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif   
14  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif   
15  Notifikasi dikembalikan ke semula saja yang ad...      5  Positif   
16  Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif   
17  Saya member platinum shopee, tp akhir akhir in...      5  Positif   
18  TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif   
19  Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif   
20  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif   
21  ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif   
22  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif   
23  Belanja online ya cm di shopee...sangat mudah ...      5  Positif   
24  Saya kasih 🌟 dulu, soalnya spaylater & spinjam...      1  Negatif   
25  Sistem pengiriman lebih sering memakan waktu l...      2  Negatif   
26  Aplikasi penipuan, baru 1 jam check out, karen...      1  Negatif   
28  Saya sangat puas belanja di shoope,, barang se...      5  Positif   
29  Shopee tambah tidak bagus dulu bisa milih jasa...      1  Negatif   
30  sekarang di shopee banyak penipu, penjual tida...      1  Negatif   
31  Apa guna nya jadi member platinum kalo paket m...      1  Negatif   
32  Makin kesini makin berat nih apk nya, terjadi ...      1  Negatif   
33  Terima kasih Shopee, kasus permasalahan saya s...      4  Positif   
34  Bayar belum jatuh tempo sudah di bayar ( dan s...      1  Negatif   
35  SEMOGA DITANGGAPI : Mohon untuk lebih selektip...      5  Positif   
37  Update terbaru sudah! Jaringan full 4G lte! Cl...      1  Negatif   
38  Tolong kembalikan kebebasan memilih jasa kirim...      1  Negatif   
39  shopee sekarang meskipun ada vocher kaya nggad...      2  Negatif   
40  Makin lama makin parah makin pelit, biaya admi...      1  Negatif   
41  Katanya terus belanja pakai spaylater supaya b...      1  Negatif   
42  Minta batal pesanan ditolak sama toko. Barang ...      1  Negatif   
43  Hallo pihak shopee, saya suka belanja online m...      4  Positif   
45  Belanja di shopee banyak variasi produknya, ku...      5  Positif   
46  Bisa ngga iklannya ga sampe mengganggu orang g...      1  Negatif   
47  kembalikan fitur yg bisa ganti ekspedisi soaln...      1  Negatif   
48  Masyallah suka sekali dengan aplikasi ini seja...      5  Positif   
49  tolong dibantu ya pihak shopee, saya setiap pe...      2  Negatif   
51  Kenapa Shopee sudah semakin berat loadingnya l...      4  Positif   
52  Jadi sekarang shopee lebih buruk dari yang lai...      1  Negatif   
54  Awal dulu bintang 5 trs krn suatu hal yg menur...      5  Positif   
55  Dulu saya kasih bintang 5 karena emang memperm...      1  Negatif   
56  Bingung aku sama shope skrng jasa kirim nya pa...      1  Negatif   
57  TOP bangeet .. Udah lama menggunakan aplikasi ...      5  Positif   

                                           text_clean  \
0   duhapa lagi ini mkin d upgrade mkin parah pela...   
1   kecewa makin lama pelayanan buruk apa lgi jasa...   
2   suka banget dengan upgrade pembatalan pembelia...   
3   shopee makin lama aja proses pengirimannya dul...   
4   sedikit kesal karna g pesan baran

# Tokenizing
Tokenizing adalah proses pemisahan teks menjadi potongan-potongan yang disebut sebagai token untuk kemudian di analisa. Kata, angka, simbol, tanda baca dan entitas penting lainnya dapat dianggap sebagai token. Didalam NLP, token diartikan sebagai “kata” meskipun tokenize juga dapat dilakukan pada paragraf maupun kalimat

In [27]:
import nltk
nltk.download('punkt')
from nltk.tokenize import sent_tokenize, word_tokenize
data_clean['text_tokens'] = data_clean['text_StopWord'].apply(lambda x: word_tokenize(x))
data_clean.head()


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


content  score    Label  \
0  duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif   
1  Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif   
2  suka banget dengan upgrade pembatalan pembelia...      1  Negatif   
3  shopee makin lama aja proses pengirimannya. Du...      1  Negatif   
4  Sedikit kesal, karna g pesan barang malah ada ...      4  Positif   

                                          text_clean  \
0  duhapa lagi ini mkin d upgrade mkin parah pela...   
1  kecewa makin lama pelayanan buruk apa lgi jasa...   
2  suka banget dengan upgrade pembatalan pembelia...   
3  shopee makin lama aja proses pengirimannya dul...   
4  sedikit kesal karna g pesan barang malah ada p...   

                                       text_StopWord  \
0  duhapa mkin d upgrade mkin parah pelayanansmen...   
1  kecewa pelayanan buruk lgi jasa pengirimnya dr...   
2  suka banget upgrade pembatalan pembelian yg kl...   
3  shopee aja proses pengirimannya mah tgl pesen ...   
4  kesal karna g pesan barang toko yg kirim baran...   

                                         text_tokens  
0  [duhapa, mkin, d, upgrade, mkin, parah, pelaya...  
1  [kecewa, pelayanan, buruk, lgi, jasa, pengirim...  
2  [suka, banget, upgrade, pembatalan, pembelian,...  
3  [shopee, aja, proses, pengirimannya, mah, tgl,...  
4  [kesal, karna, g, pesan, barang, toko, yg, kir...

# Stemming
Stemming adalah proses pemetaan dan penguraian bentuk dari suatu kata menjadi bentuk kata dasarnya. Untuk melakukan stemming bahasa Indonesia kita dapat menggunakan library Python Sastrawi yang sudah kita siapkan di awal. Library Sastrawi menerapkan Algoritma Nazief dan Adriani dalam melakukan stemming bahasa Indonesia.

In [28]:
!pip install Sastrawi

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.7/209.7 kB 5.1 MB/s eta 0:00:00


In [29]:
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
factory = StemmerFactory()
stemmer = factory.create_stemmer()

In [30]:
#-----------------STEMMING -----------------
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
#import swifter


# create stemmer
factory = StemmerFactory()
stemmer = factory.create_stemmer()

# stemmed
def stemmed_wrapper(term):
    return stemmer.stem(term)

term_dict = {}
hitung=0

for document in data_clean['text_tokens']:
    for term in document:
        if term not in term_dict:
            term_dict[term] = ' '

print(len(term_dict))
print("------------------------")
for term in term_dict:
    term_dict[term] = stemmed_wrapper(term)
    hitung+=1
    print(hitung,":",term,":" ,term_dict[term])

print(term_dict)
print("------------------------")

# apply stemmed term to dataframe
def get_stemmed_term(document):
    return [term_dict[term] for term in document]


#script ini bisa dipisah dari eksekusinya setelah pembacaaan term selesai
data_clean['text_steamindo'] = data_clean['text_tokens'].apply(lambda x:' '.join(get_stemmed_term(x)))
data_clean.head(20)

4675
------------------------
1 : duhapa : duhapa
2 : mkin : mkin
3 : d : d
4 : upgrade : upgrade
5 : parah : parah
6 : pelayanansmenjak : pelayanansmenjak
7 : expedisi : expedisi
8 : gak : gak
9 : gantipaket : gantipaket
10 : rawan : rawan
11 : bgt : bgt
12 : hilang : hilang
13 : ajapket : ajapket
14 : x : x
15 : spx : spx
16 : status : status
17 : brg : brg
18 : kota : kota
19 : tujuan : tuju
20 : tolong : tolong
21 : perhatikan : perhati
22 : pekrja : pekrja
23 : byk : byk
24 : oknum : oknum
25 : yg : yg
26 : nakal : nakal
27 : n : n
28 : bermain : main
29 : eventsoerti : eventsoerti
30 : nya : nya
31 : langganan : langgan
32 : event : event
33 : pket : pket
34 : hilangmgkn : hilangmgkn
35 : mreka : mreka
36 : barang : barang
37 : hasil : hasil
38 : co : co
39 : promo : promo
40 : disc : disc
41 : gedeanbner : gedeanbner
42 : ngerti : ngerti
43 : suara : suara
44 : dgar : dgar
45 : kecewa : kecewa
46 : pelayanan : layan
47 : buruk : buruk
48 : lgi : lgi
49 : jasa : jasa
50 : pengiri

content  score    Label  \
0   duh..apa lagi ini ..mkin d upgrade .mkin parah...      1  Negatif   
1   Kecewa makin lama Pelayanan buruk apa lgi jasa...      1  Negatif   
2   suka banget dengan upgrade pembatalan pembelia...      1  Negatif   
3   shopee makin lama aja proses pengirimannya. Du...      1  Negatif   
4   Sedikit kesal, karna g pesan barang malah ada ...      4  Positif   
5   Saya kecewa selama ini download toko online la...      5  Positif   
6   Asli aplikasi lemot banget, mau ganti nomor aj...      4  Positif   
7   Kalo aku kasih bintang 5 pasti ga dibaca. Kalo...      1  Negatif   
8   Aplikasi shopee sangat memberikan kemudahan ba...      5  Positif   
9   Saya baru tau ada toko megic ataupun penipuan,...      5  Positif   
12  Aplikasi nya udah bagus,cuman yg bikin kesal i...      1  Negatif   
14  expedisi nya skrg sdh cukup cepat sekali waktu...      5  Positif   
15  Notifikasi dikembalikan ke semula saja yang ad...      5  Positif   
16  Semenjak pihak shopee monopoli pengiriman, oto...      1  Negatif   
17  Saya member platinum shopee, tp akhir akhir in...      5  Positif   
18  TOLONG HAPUS ANTER AJA DARI LIST JASA PENGIRIM...      1  Negatif   
19  Dengan hadirnya SPX belanja di shopi sudah tid...      2  Negatif   
20  Aplikasi nya berat ,,kurang responsif Mending ...      5  Positif   
21  ekspedisinya di pilihin langsung dari Shopee, ...      1  Negatif   
22  terlalu banyak pesan notif gak jelas ,terlalu ...      1  Negatif   

                                           text_clean  \
0   duhapa lagi ini mkin d upgrade mkin parah pela...   
1   kecewa makin lama pelayanan buruk apa lgi jasa...   
2   suka banget dengan upgrade pembatalan pembelia...   
3   shopee makin lama aja proses pengirimannya dul...   
4   sedikit kesal karna g pesan barang malah ada p...   
5   saya kecewa selama ini download toko online la...   
6   asli aplikasi lemot banget mau ganti nomor aja...   
7   kalo aku kasih bintang  pasti ga dibaca kalo t...   
8   aplikasi shopee sangat memberikan kemudahan ba...   
9   saya baru tau ada toko megic ataupun penipuan ...   
12  aplikasi nya udah baguscuman yg bikin kesal it...   
14  expedisi nya skrg sdh cukup cepat sekali waktu...   
15  notifikasi dikembalikan ke semula saja yang ad...   
16  semenjak pihak shopee monopoli pengiriman otom...   
17  saya member platinum shopee tp akhir akhir ini...   
18  tolong hapus anter aja dari list jasa pengirim...   
19  dengan hadirnya spx belanja di shopi sudah tid...   
20  aplikasi nya berat kurang responsif mending il...   
21  ekspedisinya di pilihin langsung dari shopee k...   
22  terlalu banyak pesan notif gak jelas terlalu m...   

                                        text_StopWord  \
0   duhapa mkin d upgrade mkin parah pelayanansmen...   
1   kecewa pelayanan buruk lgi jasa pengirimnya dr...   
2   suka banget upgrade pembatalan pembelian yg kl...   
3   shopee aja proses pengirimannya mah tgl pesen ...   
4   kesal karna g pesan barang toko yg kirim baran...   
5   kecewa download toko online penipuanmahaldan k...   
6   asli aplikasi lemot banget ganti nomor aja nun...   
7   kalo kasih bintang ga dibaca kalo cs robot yg ...   
8   aplikasi shopee kemudahan berbelanja online fi...   
9   tau toko megic penipuan oh banget tolong ya sh...   
12  aplikasi nya udah baguscuman yg bikin kesal sp...   
14  expedisi nya skrg sdh cepat pengiriman yg defa...   
15  notifikasi dikembalikan simbol titik menarik n...   
16  semenjak shopee monopoli pengiriman otomatis p...   
17  member platinum shopee tp kecewa shoppe meruba...   
18  tolong hapus anter aja list jasa pengirimaneks...   
19  hadirnya spx belanja shopi asik lg yg pulau pa...   
20  aplikasi nya berat responsif mending ilangin s...   
21  ekspedisinya pilihin langsung shopee kalo pake...   
22  pesan notif gak mengadaada notifikasi promo pa...   

                                          text_tokens  \
0   [duhapa, mkin, d, upgrade, mkin, parah, pelaya...   
1 

In [31]:
data_clean.to_csv('hasil_TextPreProcessing_shopee.csv', index= False) #kemudian simpan hasil text preprocessing ke file csv

In [32]:
#praprocessing proses dengan menggunakan library re
import re
def praproses(text):
    text = re.sub('<[^>]*>', '', text)
    emoticons = re.findall('(?::|;|=)()(?:-)?(?:\)|\(|D|P)',
                           text)
    text = (re.sub('[\W]+', ' ', text.lower()) +
            ' '.join(emoticons).replace('-', ''))
    return text

spliting data

memecah data test 20% dari keseluruhan data

In [33]:
#membagi data menjadi data training dan testing dengan test_size = 0.20 dan random state nya 0
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_clean['content'], data_clean['Label'],
                                                    test_size = 0.20,
                                                    random_state = 0)

pembobotan tf-idf

In [34]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer()
tfidf_train = tfidf_vectorizer.fit_transform(X_train)
tfidf_test = tfidf_vectorizer.transform(X_test)


In [35]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(683,)
(683,)
(171,)
(171,)


In [36]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
vectorizer.fit(X_train)

CountVectorizer()

In [37]:
X_train = vectorizer.transform(X_train)
X_test = vectorizer.transform(X_test)

In [38]:
from sklearn.naive_bayes import MultinomialNB

nb = MultinomialNB()
nb.fit(tfidf_train, y_train)


MultinomialNB()

In [39]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [40]:
y_pred = nb.predict(tfidf_test)


In [41]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)


In [43]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

clf = MultinomialNB()
clf.fit(X_train, y_train)
predicted = clf.predict(X_test)

print("MultinomialNB Accuracy:", accuracy_score(y_test,predicted))
print("MultinomialNB Precision:", precision_score(y_test,predicted, average="binary", pos_label="Negatif"))
print("MultinomialNB Recall:", recall_score(y_test,predicted, average="binary", pos_label="Negatif"))
print("MultinomialNB f1_score:", f1_score(y_test,predicted, average="binary", pos_label="Negatif"))

print(f'confusion_matrix:\n {confusion_matrix(y_test, predicted)}')
print('====================================================\n')
print(classification_report(y_test, predicted, zero_division=0))

# Load dataset
data_clean = pd.read_csv('hasil_TextPreProcessing_shopee.csv')

MultinomialNB Accuracy: 0.7953216374269005
MultinomialNB Precision: 0.7874015748031497
MultinomialNB Recall: 0.9259259259259259
MultinomialNB f1_score: 0.851063829787234
confusion_matrix:
 [[100   8]
 [ 27  36]]

              precision    recall  f1-score   support

     Negatif       0.79      0.93      0.85       108
     Positif       0.82      0.57      0.67        63

    accuracy                           0.80       171
   macro avg       0.80      0.75      0.76       171
weighted avg       0.80      0.80      0.79       171



Kesimpulan dari evaluasi model Multinomial Naive Bayes (MultinomialNB) adalah sebagai berikut:

Akurasi: Model memiliki akurasi sebesar 0.80 atau 79.53%. Ini berarti model mampu memprediksi dengan benar sekitar 80% dari keseluruhan data uji.

Precision:

Negatif: Precision untuk kelas negatif adalah 0.79, yang berarti 79% dari prediksi negatif benar-benar negatif.
Positif: Precision untuk kelas positif adalah 0.82, yang menunjukkan bahwa 82% dari prediksi positif benar-benar positif.
Recall:

Negatif: Recall untuk kelas negatif adalah 0.93, yang berarti model berhasil mengidentifikasi 93% dari total sampel negatif yang benar.
Positif: Recall untuk kelas positif hanya 0.57, yang menunjukkan bahwa model hanya mampu mendeteksi 57% dari sampel positif yang sebenarnya.
F1-Score:

Negatif: F1-Score untuk kelas negatif adalah 0.85, mencerminkan keseimbangan yang baik antara precision dan recall.
Positif: F1-Score untuk kelas positif adalah 0.67, yang lebih rendah karena recall untuk kelas positif rendah.
Confusion Matrix:

True Negatives (100): Sebanyak 100 sampel negatif diprediksi dengan benar sebagai negatif.
False Positives (8): Sebanyak 8 sampel yang sebenarnya negatif diprediksi sebagai positif.
False Negatives (27): Sebanyak 27 sampel yang sebenarnya positif diprediksi sebagai negatif.
True Positives (36): Sebanyak 36 sampel positif diprediksi dengan benar sebagai positif.
Rata-rata (Macro Avg dan Weighted Avg):

Macro Avg: Precision, recall, dan f1-score rata-rata antar kelas masing-masing adalah 0.80, 0.75, dan 0.76. Ini menunjukkan bahwa terdapat ketidakseimbangan antara performa untuk kelas negatif dan positif.
Weighted Avg: Nilai weighted avg untuk precision, recall, dan f1-score semuanya sekitar 0.79-0.80, yang menunjukkan bahwa performa model cukup seimbang, tetapi masih lebih baik dalam memprediksi kelas negatif daripada kelas positif.
Kesimpulan Utama:
Model Multinomial Naive Bayes menunjukkan performa yang baik secara keseluruhan, terutama dalam memprediksi kelas negatif dengan recall yang tinggi. Namun, model memiliki kelemahan dalam mengidentifikasi sampel positif, yang tercermin dalam nilai recall yang rendah untuk kelas positif (57%). Model cenderung lebih baik dalam memprediksi kelas negatif dibandingkan kelas positif.










